Refactor tooltips

curioussavage opened the pull request:
This refactor should eliminate the odd/undesired behaviors of the old tooltips. 
fixes #129 and #125
Major changes:
1. There is only one tooltip that I just move around and replace the content.
2. A new cache for the responses from the tooltip endpoint.
3. tips are positioned more like traditional tooltips (the don't move with the mouse).

![valadoc](https://user-images.githubusercontent.com/7064600/34510345-fee7d47c-f007-11e7-9f1e-64eebae160d2.gif)

benwaffle commented:
For the class hierarchies the tooltip is placed at above the top of the image. Can we place it next to the currently-hovered box?

Bat41 commented on data/scripts/valadoc.js:
```diff
@@ -6,9 +6,12 @@ const config = {
 const curpkg = window.location.pathname.split('/')[1]
 
+// cache of the content of tooltips stored by their url
+var tooltipCache = {}
```
Prefer `const`.

Bat41 commented on data/scripts/valadoc.js:
```diff
@@ -6,9 +6,12 @@ const config = {
 // This object will contain the html elements of the interface
 const html = {
-  tooltips: []
+  tooltipEl: initTooltip(),
```
Not sure about the trailing comma.

curioussavage replied:
I prefer it for cleaner diffs in the future but I'm happy to change it to whatever you prefer.

Bat41 commented on data/scripts/valadoc.js:
```diff
@@ -34,50 +37,58 @@ function search (query) {
+    const tipRect = tip.getBoundingClientRect()
+    const tipOffset = 5 + tipRect.height
+    tip.style.top = (targetRect.top + pageYOffset - tipOffset) + 'px'
+    tip.style.left = (targetRect.x + pageXOffset) + 'px'
```
Maybe using format strings would be nicer?

curioussavage commented:
@Bat41 I fixed those es6 issues and went ahead and rebased on master.

@benwaffle could you give some more clarification on your request? Tooltips are typically above the item hovered.

running on ubuntu 16 based distro here php 7 installed and commit 2d5cd3cc0ee2dbbf1ebf9ea9268bb02f66228927 appears to have broken my server.

![image](https://user-images.githubusercontent.com/7064600/34547754-5161012e-f0b2-11e7-9ba0-acad28cf3301.png)

benwaffle commented:
@curioussavage 
![bug](https://user-images.githubusercontent.com/1713819/34550452-e82242d4-f108-11e7-9320-41fd2f8773d3.gif)

curioussavage commented:
thanks. Will have to look at it tomorrow morning.

curioussavage commented:
![class heirarchy diagram](https://user-images.githubusercontent.com/7064600/34637609-9e63329a-f26e-11e7-8111-de44795deb85.gif)

Do you guys like them aligned to the image? I like it because then going up and down the hierarchy it wont jump left to right.

benwaffle commented:
Pretty sure your router issues are due to this https://github.com/php/php-src/commit/816758eda2bcdd69ba505fb6bbb79124a7bf2254

benwaffle commented:
When the SVG is under the top bar, the tooltip is incorrectly positioned. ![screenshot from 2018-01-07 01-09-02](https://user-images.githubusercontent.com/1713819/34647142-42032794-f372-11e7-8f96-50c0a57a111a.png)

And i'm not sure that always putting it on the right will be OK with large class heirarchies:
![screenshot from 2018-01-07 01-17-05](https://user-images.githubusercontent.com/1713819/34647155-9bcbf332-f372-11e7-97e8-b3f42314f21a.png)

curioussavage commented:
> And i'm not sure that always putting it on the right will be OK with large class heirarchies:

In that case should I just always position on top? I personally like that more for consistency.

ZanderBrown commented:
I'm not sure why your hard aligning them to one side anyway. What's your reasoning?

benwaffle commented:
I think that putting the tooltip next to each box is best. I can work on this in a couple hours, if you don't do it first.

curioussavage commented:
I fixed the issue with the vertical alignment for tooltips on area tags. 

There is an annoying bug that happens when you hover over a tooltip with an area tag as target and then a normal link. For some reason it gets the old height for the div. This only occurs when you go from an `area -> a`. I double checked many times and when moving from a link to a link the height is the updated new height.

benwaffle commented:
congratulations!

benwaffle commented:
I think this looks better
![out](https://user-images.githubusercontent.com/1713819/34704006-1de0a23a-f4c4-11e7-85b8-4d51860765de.gif)
